test/integration/src/test/java/io/pravega/test/integration/ReaderGroupNotificationTest.java

```diff
@@ -189,24 +196,107 @@ public void testSegmentNotifications() throws Exception {
+        EventStreamWriter<String> writer1 = clientFactory.createEventWriter(streamName, new JavaSerializer<>(),
+            EventWriterConfig.builder().build());
+        List<String> routingKeys = new ArrayList<>();
+        routingKeys.add("0");
+        @Cleanup
+        ReaderGroupManager groupManager = new ReaderGroupManagerImpl(SCOPE, controller, clientFactory,
+            connectionFactory);
+        groupManager.createReaderGroup("reader", ReaderGroupConfig
+            .builder().disableAutomaticCheckpoints().stream(Stream.of(SCOPE, streamName)).groupRefreshTimeMillis(0).build());
+        @Cleanup
+        ReaderGroup readerGroup = groupManager.getReaderGroup("reader");
+        @Cleanup
+        EventStreamReader<String> reader1 = clientFactory.createReader("readerId", "reader", new JavaSerializer<>(),
+            ReaderConfig.builder().initialAllocationDelay(0).build());
+
+        val notificationResults = new ArrayBlockingQueue<SegmentNotification>(10);
+
+        //Add segment event listener
+        Listener<SegmentNotification> l1 = notification -> {
+            log.info("Number of Segments: {}, Number of Readers: {}", notification.getNumOfSegments(), notification.getNumOfReaders());
+            notificationResults.add(notification);
+        };
+        SegmentNotifier segmentNotifier = (SegmentNotifier) readerGroup.getSegmentNotifier(executor);
+        segmentNotifier.registerListener(l1);
+
+        long start = System.currentTimeMillis();
+        CompletableFuture.runAsync(() -> {
+            while (System.currentTimeMillis() - start < Duration.ofMinutes(3).toMillis()) {
+                try {
+                    writer1.writeEvent(routingKeys.get(0), "data1").get();
+                } catch (Throwable e) {
+                    System.err.println("test exception writing events " + e.getMessage());
+                    break;
+                }
+            }
+        });
+
+        Retry.withExpBackoff(10, 10, 100, 10000)
+            .retryingOn(IllegalStateException.class)
+            .throwingOn(RuntimeException.class)
+            .runAsync(() -> controller.getCurrentSegments(SCOPE, streamName)
+                .thenAccept(streamSegments -> {
+                    log.info("segment_size={}", streamSegments.getSegments().size());
+                    segmentNotifier.pollNow();
+                    SegmentNotification initialSegmentNotification = null;
+                    try {
+                        initialSegmentNotification = notificationResults.take();
+                    } catch (InterruptedException e) {
+                        log.error("Exception:", e);
+                        throw new IllegalStateException(e);
+                    }
+                    assertNotNull(initialSegmentNotification);
+                    assertTrue(initialSegmentNotification.getNumOfSegments() > 2);
+                    if (streamSegments.getSegments().size() > 2) {
```

Comment:
if the initial number of segments is 3 then the test would check the currentSegments once and would be successful.